spark/spark-tensorflow-distributor/spark_tensorflow_distributor/mirrored_strategy_runner.py

```diff
@@ -0,0 +1,331 @@
+            with gpu_resource_name='gpu' would be `{"name": "gpu", "addresses":["0","1","2"]}`. See an example discovery
+            script: `github.com/apache/spark/blob/master/examples/src/main/scripts/getGpusResources.sh`.
+        :param use_custom_strategy: When true, the training function passed to the MirroredStrategyRunner.run method must
+            construct and use its own tensorflow.distribute.Strategy() object. When false, MirroredStrategyRunner constructs
+            one for the user and wraps the training function in the strategy context, allowing the user to provide
+            non-distributed TensorFlow code that is executed as distributed code.
+
+            Example with use_custom_strategy=True:
+
+                def train_fn():
+                    import tensorflow as tf
+                    strategy = tf.distribute.experimental.MultiWorkerMirroredStrategy()
+                    with strategy.scope():
+                        # training code
+
+            Example with use_custom_strategy=False:
+
+                def train_fn():
+                    import tensorflow as tf
+                    # training code
+        """
+        self.logger = _get_logger(self.__class__.__name__)
+        self.num_slots = num_slots
+        if num_slots == 0:
+            raise ValueError(
+                'num_slots cannot be 0.'
+            )
+        self.use_custom_strategy = use_custom_strategy
+        self.gpu_resource_name = gpu_resource_name
+        self.is_gpu = self.gpu_resource_name is not None
+        if self.is_gpu:
+            self.logger.info(
+                'Doing GPU training...'
+            )
+        else:
+            self.logger.info(
+                'Doing CPU training...'
+            )
+        self.local_mode = self.num_slots < 0
+        spark = SparkSession.builder.getOrCreate()
+        self.sc = spark.sparkContext
+        if self.local_mode is True:
+            self.logger.warning(
+                'MirroredStrategyRunner will run in local mode on the driver node. '
+                'There would be resource contention if the driver also runs other workloads.'
+            )
+            self.num_slots = -self.num_slots
+            self.num_tasks = None
+            self.num_workers = None
+        else:
+            self.num_tasks = self.get_num_tasks()
+            self.logger.info(f'Will run with {self.num_tasks} Spark tasks.')
+
+    def get_num_tasks(self):
+        if self.is_gpu:
+            key = f'spark.task.resource.{self.gpu_resource_name}.amount'
+            if not self.sc.getConf().contains(key):
+                raise Exception(
+                    'Your cluster might not have Spark GPU-aware scheduling enabled, '
+                    'please contact your cluster administrator.'
+                    f'The conf `{key}` was not found in the Spark configuration.'
+                )
+            task_gpu_amount = int(self.sc.getConf().get(key))
```

Comment:
Is it always True that in a spark cluster all workers have the same number of GPUs?

Reply:
No, but Spark 3.0's resource aware scheduling (used here) guarantees that each Spark task is allocated the same number of GPUs. This value is specified by the cluster admin in the Spark conf `spark.task.resource.gpu.amount` and the task program can find which GPUs it is allocated by Spark with the `BarrierTaskContext.resources()` method.